distributed/shuffle/tests/test_shuffle.py

```diff
@@ -124,6 +135,32 @@ async def test_basic_integration(c, s, a, b, lose_annotations):
     await clean_scheduler(s)
 
 
+@pytest.mark.parametrize("npartitions", [None, 1, 20])
+@gen_cluster(client=True)
+async def test_shuffle_with_array_conversion(c, s, a, b, lose_annotations, npartitions):
+    await invoke_annotation_chaos(lose_annotations, c)
+    df = dask.datasets.timeseries(
+        start="2000-01-01",
+        end="2000-01-10",
+        dtypes={"x": float, "y": float},
+        freq="10 s",
+    )
+    out = dd.shuffle.shuffle(df, "x", shuffle="p2p", npartitions=npartitions).values
+
+    if npartitions == 1:
+        # FIXME: distributed#7816
+        with raises_with_cause(
+            RuntimeError, "shuffle_transfer failed", RuntimeError, "Barrier task"
+        ):
+            await c.compute(out)
+    else:
+        await c.compute(out)
+
+    await clean_worker(a)
+    await clean_worker(b)
+    await clean_scheduler(s)
```

Comment:
Not relevant to your PR, but I found these to be strangely named functions. On the surface I expected them to be some sort of, well, cleanup functions. Turns out they're asserting no shuffle state, So would think `assert no_shuffle_state(...)`

Reply:
Good point! I encourage you to dig deeper into the P2P code. I suspect that you will find many things that don't make sense to you and that I'll need to clean up and document. Helping me find those would be greatly appreciated :)